[FSDP] Add `scatter_full_optim_state_dict()`

awgu opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #75603 [FSDP] `scatter_full_optim_state_dict()`: async broadcast
* **#75517 [FSDP] Add `scatter_full_optim_state_dict()`**

**Overview**
This adds `scatter_full_optim_state_dict()`, which is an alternate way to load an optimizer state checkpoint. It has the same effect as `broadcast_object_list([full_optim_state_dict], src=0)` + `shard_full_optim_state_dict(full_optim_state_dict, model)` (on all ranks).
- All ranks call the method with rank 0 passing in the populated full optimizer state dict (returned by `full_optim_state_dict()`).
- Rank 0 first broadcasts the state dict with positive-dimension tensor states replaced with their metadata and then broadcasts the actual tensor states. For tensor states corresponding to sharded FSDP parameters, ranks only save the relevant shard.

Differential Revision: [D35550062](https://our.internmc.facebook.com/intern/diff/D35550062)

Commit:
[FSDP] Add `scatter_full_optim_state_dict()`

[ghstack-poisoned]

Commit:
Update on "[FSDP] Add `scatter_full_optim_state_dict()`"




[ghstack-poisoned]

awgu commented:
@awgu has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D35550062).

awgu commented:
To-do: Change `broadcast()`s to async, save handles, and wait on them.

Commit:
Update on "[FSDP] Add `scatter_full_optim_state_dict()`"


**Overview**
This adds `scatter_full_optim_state_dict()`, which is an alternate way to load an optimizer state checkpoint. It has the same effect as `broadcast_object_list([full_optim_state_dict], src=0)` + `shard_full_optim_state_dict(full_optim_state_dict, model)` (on all ranks).
- All ranks call the method with rank 0 passing in the populated full optimizer state dict (returned by `full_optim_state_dict()`).
- Rank 0 first broadcasts the state dict with positive-dimension tensor states replaced with their metadata and then broadcasts the actual tensor states. For tensor states corresponding to sharded FSDP parameters, ranks only save the relevant shard.

Differential Revision: [D35550062](https://our.internmc.facebook.com/intern/diff/D35550062)

[ghstack-poisoned]

awgu commented:
@awgu has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D35550062).

Commit:
Update on "[FSDP] Add `scatter_full_optim_state_dict()`"


**Overview**
This adds `scatter_full_optim_state_dict()`, which is an alternate way to load an optimizer state checkpoint. It has the same effect as `broadcast_object_list([full_optim_state_dict], src=0)` + `shard_full_optim_state_dict(full_optim_state_dict, model)` (on all ranks).
- All ranks call the method with rank 0 passing in the populated full optimizer state dict (returned by `full_optim_state_dict()`).
- Rank 0 first broadcasts the state dict with positive-dimension tensor states replaced with their metadata and then broadcasts the actual tensor states. For tensor states corresponding to sharded FSDP parameters, ranks only save the relevant shard.

Differential Revision: [D35550062](https://our.internmc.facebook.com/intern/diff/D35550062)

[ghstack-poisoned]

Commit:
Update on "[FSDP] Add `scatter_full_optim_state_dict()`"


**Overview**
This adds `scatter_full_optim_state_dict()`, which is an alternate way to load an optimizer state checkpoint. It has the same effect as `broadcast_object_list([full_optim_state_dict], src=0)` + `shard_full_optim_state_dict(full_optim_state_dict, model)` (on all ranks).
- All ranks call the method with rank 0 passing in the populated full optimizer state dict (returned by `full_optim_state_dict()`).
- Rank 0 first broadcasts the state dict with positive-dimension tensor states replaced with their metadata and then broadcasts the actual tensor states. For tensor states corresponding to sharded FSDP parameters, ranks only save the relevant shard.

Differential Revision: [D35550062](https://our.internmc.facebook.com/intern/diff/D35550062)

[ghstack-poisoned]

Commit:
Update on "[FSDP] Add `scatter_full_optim_state_dict()`"


**Overview**
This adds `scatter_full_optim_state_dict()`, which is an alternate way to load an optimizer state checkpoint. It has the same effect as `broadcast_object_list([full_optim_state_dict], src=0)` + `shard_full_optim_state_dict(full_optim_state_dict, model)` (on all ranks).
- All ranks call the method with rank 0 passing in the populated full optimizer state dict (returned by `full_optim_state_dict()`).
- Rank 0 first broadcasts the state dict with positive-dimension tensor states replaced with their metadata and then broadcasts the actual tensor states. For tensor states corresponding to sharded FSDP parameters, ranks only save the relevant shard.

Differential Revision: [D35550062](https://our.internmc.facebook.com/intern/diff/D35550062)

[ghstack-poisoned]

awgu commented:
@awgu has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D35550062).

awgu commented on test/distributed/fsdp/test_fsdp_optim_state.py:
```diff
@@ -433,39 +486,56 @@ def _test_shard_full_optim_state(
-        # local optimizer's state dict
+        # Compute two sharded optim state dicts: (1) for the first model
+        # according to the second model and (2) for the second model according
+        # to the second model
```
We see below how calling `FSDP.scatter_full_optim_state_dict()` is equivalent to broadcasting the full optimizer state dict and calling `FSDP.shard_full_optim_state_dict()`.

In other words, `sharded_osd1` and `sharded_osd2` are the same regardless of `osd_comm_method` (except for possibly the tensors' residing device).

awgu commented on test/distributed/fsdp/test_fsdp_optim_state.py:
```diff
@@ -433,39 +486,56 @@ def _test_shard_full_optim_state(
+                full_osd2 if self.rank == 0 else None, model2, optim_input2,
+                group=new_group,
+            )
+            self._check_state_device(sharded_osd1, on_gpu=True)
```
It makes sense to ensure tensors to be on GPU for `scatter_full_optim_state_dict()` since the positive-dimension tensors will already be on GPU after `broadcast()`.

However, I have not decided whether to move tensors to GPU as a part of `shard_full_optim_state_dict()` or to leave that up to the user. The tensors in the dict returned by `full_optim_state_dict()`, which is the assumed first argument to `shard_full_optim_state_dict()`, are on CPU.
What are your thoughts? Should we build the CPU -> GPU transfer into the API or leave it to the user?

awgu replied:
I will add a follow-up to add a device option for `shard_full_optim_state_dict()`.

awgu commented on torch/distributed/fsdp/fully_sharded_data_parallel.py:
```diff
@@ -727,32 +730,60 @@ def _shard_parameters(self) -> None:
-        # Shard using torch.chunk to match all-gather/reduce-scatter.
-        chunks = torch.flatten(tensor).chunk(self.world_size)
-        if len(chunks) < (self.rank + 1):
+    @staticmethod
```
`_get_chunk()` is separated since it includes all the logic up to but not including the point where new tensor storage may be allocated. This is needed to compute the sharded tensor shape without allocating new storage.

awgu commented on torch/distributed/fsdp/fully_sharded_data_parallel.py:
```diff
@@ -727,32 +730,60 @@ def _shard_parameters(self) -> None:
-        # We always need to clone here, because regardless of padding the
-        # original parameter, of which this chunk is a view of, is deallocated
-        # after _get_shard.
+    @staticmethod
```
`_get_shard_functional()` is needed to compute the shard shape for a target rank that may not be the calling rank.

awgu commented on torch/distributed/fsdp/fully_sharded_data_parallel.py:
```diff
@@ -2539,70 +2569,109 @@ def shard_full_optim_state_dict(
-                "`full_optim_state_dict` must have the keys \"state\" and "
-                "\"param_groups\" to be a valid optimizer state dict"
-            )
+        return _flatten_full_optim_state_dict(
```
The flattening logic for `shard_full_optim_state_dict()` and for `scatter_full_optim_state_dict()` on rank 0 are the same, except for the former, only the shard relevant to the rank is kept, while for the latter, each shard is kept, to be broadcast shard-by-shard later.

To this end, I refactored the common logic into `_flatten_full_optim_state_dict()` with a `bool` argument `shard_state`. It is `True` for `shard_full_optim_state_dict()` and `False` for `scatter_full_optim_state_dict()`.

awgu commented on torch/distributed/fsdp/fully_sharded_data_parallel.py:
```diff
@@ -2539,70 +2569,109 @@ def shard_full_optim_state_dict(
+            flat_osd, fsdp_flat_param_ids = _flatten_full_optim_state_dict(
+                full_optim_state_dict, model, False, optim_input,
+            )
+            processed_osd = _process_pos_dim_tensor_state(
```
This replaces all positive-dimension tensor state with metadata. This is because this `processed_osd` needs to be broadcast from rank 0 to all ranks so that each nonzero rank knows how to receive positive-dimension tensor shards. (It needs the optimizer state dict "skeleton".)

awgu commented on torch/distributed/fsdp/optim_utils.py:
```diff
@@ -192,11 +221,95 @@ def _unflatten_communicated_optim_state(
     return unflat_param_state
 
 
+def _flatten_full_optim_state_dict(
```
Common logic to both `shard_full_optim_state_dict()` and `scatter_full_optim_state_dict()`, only differentiated by `shard_state: bool`.

awgu commented on torch/distributed/fsdp/optim_utils.py:
```diff
@@ -488,16 +607,261 @@ def _flatten_non_tensor_optim_state(
+    """
+    flat_osd = flat_optim_state_dict  # alias
+    no_tensor_osd: Dict[str, Any] = {"state": {}}
+    cpu_device = torch.device("cpu")
```
To-do: looks unused now; remove.

awgu reviewed (commented):
Leaving some comments to help with review.

zhaojuanmao commented on torch/distributed/fsdp/optim_utils.py:
```diff
@@ -488,16 +607,261 @@ def _flatten_non_tensor_optim_state(
+    for target_rank in range(1, world_size):
+        if rank == 0:
+            assert get_shard is not None
+            sharded_tensor = get_shard(target_rank, world_size)[0].to(broadcast_device)
```
nit: looks like no need to call .to(broadcast_device) as unsharded_tensor has already been moved? maybe I'm wrong

awgu replied:
Good catch. I fixed this so that only the shard is moved to `broadcast_device`. Otherwise, the peak GPU memory usage on rank 0 is possibly increased.

zhaojuanmao reviewed (commented):
Thanks for adding this API, overall looks great, just some minor comments. 

Also would you please document when scatter_full_optim_state_dict() API is recommended and when shard_full_optim_state_dict() is recommend? 

My understanding is following: 
1. scatter_ API has better memory efficiency at the cost of communication cost when loading states on CPU, as CPU memory is shared by 8 ranks, saving CPU memory can help loading larger models. 
2. shard_ API is recommended for loading states on GPU memory, the largest model size can be loaded is constrained by rank 0's GPU memory size any way, scatter_ API does not really help much to load larger models. So shard_ API is recommended, as it is faster without communication cost

awgu commented on torch/distributed/fsdp/optim_utils.py:
```diff
@@ -192,11 +221,95 @@ def _unflatten_communicated_optim_state(
+            flat_osd_state[flat_param_id] = \
+                copy.copy(full_osd_state[unflat_param_name])
+            unflat_param_names_to_flat_param_id[unflat_param_name] = flat_param_id
+
```
No need since the `fsdp_flat_param_ids` is meant specifically to identify which IDs correspond to `FSDP` flattened parameters. This is to know which tensor state needs to be sharded by rank 0 and broadcast shard by shard as opposed to not sharded and broadcast as is.

Commit:
Update on "[FSDP] Add `scatter_full_optim_state_dict()`"


**Overview**
This adds `scatter_full_optim_state_dict()`, which is an alternate way to load an optimizer state checkpoint. It has the same effect as `broadcast_object_list([full_optim_state_dict], src=0)` + `shard_full_optim_state_dict(full_optim_state_dict, model)` (on all ranks).
- All ranks call the method with rank 0 passing in the populated full optimizer state dict (returned by `full_optim_state_dict()`).
- Rank 0 first broadcasts the state dict with positive-dimension tensor states replaced with their metadata and then broadcasts the actual tensor states. For tensor states corresponding to sharded FSDP parameters, ranks only save the relevant shard.

Differential Revision: [D35550062](https://our.internmc.facebook.com/intern/diff/D35550062)

[ghstack-poisoned]

awgu commented on torch/distributed/fsdp/fully_sharded_data_parallel.py:
```diff
@@ -2651,6 +2681,18 @@ def shard_full_optim_state_dict(
+            communicates it to ranks appropriately. Hence, the former has
+            higher aggregate CPU memory cost, while the latter has higher
+            communication cost.
+
```
Added a note above to explain difference between the two APIs. (replicated on both methods)

Commit:
Update on "[FSDP] Add `scatter_full_optim_state_dict()`"


**Overview**
This adds `scatter_full_optim_state_dict()`, which is an alternate way to load an optimizer state checkpoint. It has the same effect as `broadcast_object_list([full_optim_state_dict], src=0)` + `shard_full_optim_state_dict(full_optim_state_dict, model)` (on all ranks).
- All ranks call the method with rank 0 passing in the populated full optimizer state dict (returned by `full_optim_state_dict()`).
- Rank 0 first broadcasts the state dict with positive-dimension tensor states replaced with their metadata and then broadcasts the actual tensor states. For tensor states corresponding to sharded FSDP parameters, ranks only save the relevant shard.

Differential Revision: [D35550062](https://our.internmc.facebook.com/intern/diff/D35550062)

[ghstack-poisoned]

awgu commented:
Made `optim_utils.py` private (`_optim_utils.py`). I could not figure out how to deal with the `typing` imports breaking the new public API requirements: https://github.com/pytorch/pytorch/wiki/Public-API-definition-and-documentation

awgu commented:
@awgu has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D35550062).

awgu commented:
@pytorchbot merge this please
